design-proposals/klip-1-improve-udf-interfaces.md

```diff
@@ -0,0 +1,291 @@
+@UdfReturn(file   = 'schema_def.kschema')           // another way pointing to a file
+@UdfReturn(schema = "name.space.MyClass.SCHEMA")    // another way that would resolve a java object
+public Struct generate() {
+  return new Struct(...);
+}
+```
+
+### Generics
+
+Today, we eagerly convert UDF output types to corresponding Connect schema types at the time that we
+compile the UDFs and not the time that we use them in a `SELECT` clause. Since there is no 
+corresponding "unresolved" type in Connect, we fail this conversion. To solve this problem, we can
+infer the type from the source schema and compile the code to work with any component type (e.g.
+omit the generics from the compiled code).
+
+```java
+@Udf("returns a sublist of 'list'")
+public <T> List<T> sublist(
+    @UdfParameter final List<T> list, 
+    @UdfParameter final int start, 
+    @UdfParameter final int end) {
+  return list.subList(start, end);
+}
+```
+
+We need to ensure that the output type can be inferred. This means that we need to fail if the 
+signature is something like `<T> List<T> convert(List<String> list)`. This can be done during 
+compilation by following a simple rule: any generic type used in the output must be present in at
+least one of the parameters. We can access this information via reflection.
+
+**NOTE:** Supporting a wildcard type (e.g. `<?>`) is not covered by this design since we would not
+be able to generate the output schema of select statements.
+
+### Varargs
+
+Varargs boils down to supporting native Java arrays as parameters and ensuring that component types 
+are resolved properly. Anytime a method is registered with variable args (e.g. `String... args`) we 
+will register a corresponding function with the wrapping array type (e.g. `String[]`). At runtime, 
+if no function matches the explicit parameter types, we will fallback to search for any array-type 
+parameter method declaration of the same type iff all parameters are of a single type. If any
+parameter is null, it will be considered valid for any vararg declaration.
+
+This proposal does not cover supporting `Object` as a parameter to a method in order to allow for 
+"generic variable argument" UDFs, but supporting it can be an extension.
+
+```java
+@Udf("returns a sublist of 'list' t")
+public int sum(@UdfParameter final int... args) {
+  return Arrays.stream(args).sum();
+}
+```
+
+### Complex Aggregation
+
+We will allow users to supply an additional method `VR export(A agg)`. This method will be taken
+as the behavioral parameter to a `mapValues` task that will be applied after the `aggregate` task 
+in the generated KStreams topology. To support this in a backwards compatible fashion, we will 
+introduce a new interface `Exportable`, that UDFs may implement. Only UDAFs that implement this 
+interface will have the `mapValues` stage applied to them.
+
+```java
+class AverageUdaf implements Udaf<Long, Struct>, Exportable<Struct, Long> {
```

Comment:
Few questions here:
- thinking from the perspective of someone who might implement a UDAF using this new capability, the `Udaf<Long, Struct>` part seems non-intuitive - i'd expect here to declare the output type of my UDAF (only). Feels like I'm telling you what i want to serialize between invocations in the `Exportable` declaration, and then unnecessarily repeating it in the `Udaf` part ?  (or does this have to be here for backwards compatibility reasons i'm not quite seeing?)
- perhaps another option to try to simplify the ux could be a new `ExportableUdaf` interface, thereby not requiring the second interface ?
- another random thought on backwards compatibility: we could have a default no-op implementation of the `export` method on the interface itself ? Although i guess this simplification would be traded against the fact that we'd then do an unnecessary invocation of `mapValues` for UDAFs which didn't strictly require it ? 
Apologies if those Qs are off the mark or missing something obvious, i'm very rusty!

Reply:
> * (or does this have to be here for backwards compatibility reasons i'm not quite seeing?)

The problem here, is that existing signatures would be messed up. Imagine I have the following class:
```java
class MyUdaf implements Udaf<Long, Long> {
  public Long aggregate(final Long val, final Long agg);
}
```

I want to implement the following (using my syntax):
```java
class MyExportUdaf implements Udaf<Long, Struct>, Exportable<Struct, Long> {
    public Struct aggregate(final Long val, final Struct agg);
    public Long export(Struct agg);
}
```

If I change `Udaf<V, A>` to `Udaf<V, VR>`, then `MyExportUdaf#aggregate` doesn't work:
```java
class MyExportUdaf implements Udaf<Long, Long>, Exportable<Struct, Long> {
    public Struct aggregate(final Long val, final Long agg); // don't have access to `Struct`
}
```

On the other hand, if I have `Udaf<V, A, VR>` then `MyUdaf` no longer compiles because it doesn't specify `VR`.

> * we could have a default no-op implementation of the export method on the interface itself

See above. We can't do this because we need to know the export types. If we add `VR export(A)` to `Udaf`, we need a way to define `VR` (simply adding it will break all existing Udafs).

> * perhaps another option to try to simplify the ux could be a new ExportableUdaf interface, thereby not requiring the second interface ?

Definitely a possibility, and it could `ExportableUdaf<V, A, VR> extends Udaf<V, VR>`. I actually quite like this option.